Blacklist gulp-better-rollup

Permutatrix opened the pull request:
[This plugin reads from the filesystem directly, which is what the original gulp-rollup was blacklisted for.](https://github.com/MikeKovarik/gulp-better-rollup/issues/2) gulp-rollup still doesn't show up in the plugin search (has the blacklist not been updated?), but gulp-better-rollup does, and I'm not having it.

phated commented:
Thanks for bringing this to our attention. I'll review in a day or 2. I think we need to redeploy the blacklist. Sorry for the delay.

MikeKovarik commented:
Unfortunately Rollup opens the files on its own and buffer cannot be stuffed into it and there's nothing I can do about it. And I am aware gulp plugins used before rollup wouldn't have any effect because of that. I am sorry for this situation and if it need to be blacklisted then so be it.

BUT... I took a look at what `gulp-rollup` does differently than my plugin but what i found out (i could be wrong) is it's not perfect either. It doesn't use `gulp.src` for globbing because it doesn't use it at all. It's just faking it and `gulp.src` just sits in the task with dummy path `./src/**/*.js` while actual entry file path is defined in the options object of the plugin. And to be honest that is more complicated API as well as hypocritical because like in `gulp-better-rollup` it reads (or rather rollup inside) directly from filesystem instead of using gulp pipes which leads to inability to use any plugin before rollup just like in my case.

The only difference is that `gulp-better-rollup` uses nicer syntax
```
gulp.task('bundle', function() {
  gulp.src('./src/main.js')
    .pipe(sourcemaps.init())
    .pipe(rollup())
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('./dist'));
});
```
whereas with `gulp-rollup` you need to write this
```
gulp.task('bundle', function() {
  gulp.src('./src/**/*.js')
    .pipe(sourcemaps.init())
    .pipe(rollup({
        entry: './src/main.js'
    }))
    .pipe(sourcemaps.write())
    .pipe(gulp.dest('./dist'));
});
```

So `gulp-rollup` should be blacklisted as well

contra commented:
If it isn't possible for gulp-rollup to work without accessing the fs, then it should be at the head of the stream (like how webpack and browserify work).

Example:

```js
gulp.task('bundle', function() {
  return rollup({ entry: './src/main.js' }).pipe(gulp.dest('./dist'));
});
```

Permutatrix commented:
@MikeKovarik No no no no no, that's a severe misunderstanding of how it works. The path passed into `gulp.src` for gulp-rollup is _not_ a dummy. gulp-rollup builds up a "virtual filesystem" from the files it reads from the stream&mdash;it's pure Vinyl, it doesn't access the filesystem unless you explicitly allow it to, which isn't recommended. That's why you have to pass the entry path as an option&mdash;you need to collect every file which might be imported in the stream, and most of those aren't going to be entry points.

But yes, it's a terrible API for general use. That's because gulp-rollup isn't intended for general use. It's intended for the very specific case where you have to use a Vinyl stream to _synthesize_ your files. The tool I created for general use is [rollup-stream](https://github.com/gulpjs/gulp/blob/master/docs/recipes/rollup-with-rollup-stream.md). This is explained and the tool linked to in the gulp-rollup README, and it goes at the head of the stream like @contra suggests.

contra commented:
Doesn't look like it touches the fs so I'm going to close this. Thanks for chiming in @Permutatrix 🌴

contra commented:
Unrelated, but it would be nice if you merged with gulp-rollup (or took over that package name) somehow so there aren't two - then we could remove gulp-rollup from the list.

Permutatrix commented:
@contra You have it backwards. gulp-better-rollup calls Rollup, which undoubtedly touches the filesystem. gulp-rollup _was_ blacklisted _in the past_ for this until I fixed it, so it's definitely a problem. Now that I have fixed it, it's been reviewed again and removed, so I certainly shouldn't be merging gulp-rollup-better. Please reopen this.

contra commented:
@Permutatrix Where inside gulp-better-rollup is it touching the filesystem? It looks like they're giving rollup a cache w/ the vinyl files.

Permutatrix commented:
@contra Are you looking at the right package? Because that's what gulp-rollup does, not gulp-better-rollup. You can just Ctrl+F for "file" in gulp-better-rollup's [index.js](https://github.com/MikeKovarik/gulp-better-rollup/blob/master/index.js) and you'll see that the contents are never used.

contra commented:
@Permutatrix Yep, I'm looking at gulp-better-rollup. I've never used rollup though so I don't know how it works without a base path, but I assumed giving it a cache and no base option would not read from or write to the fs.

Permutatrix commented:
@contra The cache in Rollup is used to avoid repeatedly calculating the same transformations on the same code, not to avoid filesystem accesses.

MikeKovarik commented:
@Permutatrix OK, I'm sorry for the misunderstanding. I've spent quite a while looking at `gulp-rollup` and that wasn't able to wrap my head around it. Could you please explain how exactly do you force rollup to accept the vinyl file? I'd for sure be doing that if rollup api accepted buffered input it but it only accepts [string path](https://github.com/rollup/rollup/wiki/JavaScript-API#entry) to the file. From what I'm seeing you're still passing options.entry into rollup in `gulp-rollup`

@contra It's no me who touches the fs but I'm responsible for it by providing rollup with `entry` file path in `gulp-better-rollup`. Rollup then reads the file from fs. @contra is right, cache is only there to save time by avoiding transformation on unchanged files and it stores AST or something, not the file contents.

About merging: Yeah It's certainly not ideal to have two plugins for the same thing. But `gulp-better-rollup` uses quite different approach and I didn't want to just rush into a `gulp-rollup` repo and hijack it :D. I'm also thinking of reworking it as a vinyl adapter as suggested or something else. But I want it to work with `gulp-watch` and there's quite a few things to figure out. So until then I guess there's no other way than blacklisting `gulp-better-rollup`

Side question: If there's a way (in the future) to pass vinyl file into rollup as entry, would it be ok for rollup to then access fs for the dependencies of this file? Because the point of rollup is to build a package of the entry file and all of it's deps.

Permutatrix commented:
@MikeKovarik The greater part of gulp-rollup is actually [rollup-plugin-hypothetical](https://github.com/Permutatrix/rollup-plugin-hypothetical), which is sneakily and (in all practical cases) transparently added to the array of plugins passed into Rollup. This allows it to intercept Rollup's file accesses and replace them with in-memory dictionary lookups.

> Side question: If there's a way (in the future) to pass vinyl file into rollup as entry, would it be ok for rollup to then access fs for the dependencies of this file?

That behavior was proposed for the original gulp-rollup, and I opposed it: https://github.com/mcasimir/gulp-rollup/pull/16. The fundamental flaw with the design of gulp-better-rollup and the original gulp-rollup isn't that it doesn't _use_ the data passed in to it, but that it doesn't make sense for that data to be passed in at all. After all, none of the other files' contents are being passed in. Regardless of its behavior, the plugin's API, despite looking intuitive at first glance when used in the context of a gulp task, just doesn't make sense.

> Because the point of rollup is to build a package of the entry file and all of it's deps.

Hence @contra's suggestion: https://github.com/gulpjs/plugins/pull/221#issuecomment-264661203. A typical invocation of Rollup takes in no file contents; it only outputs them. It therefore makes sense to put it at the start of the stream, where `gulp.src` would often be, and where we're allowed to do whatever we want, including reading from the filesystem. That's why I made [rollup-stream](https://github.com/Permutatrix/rollup-stream).

phated commented:
I am in agreement with @Permutatrix that this plugin is violating guidelines and the reason for it doesn't seem to make sense (watching only the entry due to globbing seems wrong, etc).  I think all legit use cases are supported by gulp-rollup and rollup-stream.

Btw, @Permutatrix, your work on rollup-plugin-hypothetical and gulp-rollup is amazing and I wish more authors did such great work.

phated commented:
@Permutatrix I've redeployed so gulp-rollup is showing on the plugins site now.  Thanks for pointing that out.
